feat(image-resolution): Implement Image resolution

tonymbugua-outreach opened the pull request:
<!--
  !!!! README !!!! Please fill this out.

  Please follow the PR naming conventions: 
  https://outreach-io.atlassian.net/wiki/spaces/EN/pages/1902444645/Conventional+Commits
-->


<!-- A short description of what your PR does and what it solves. -->
## What this PR does / why we need it
We want to use "stable" images when deploying applications within devenv (i.e. orange or green channels within maestro


<!-- <<Stencil::Block(jiraPrefix)>> -->
## Jira ID

DT-3506
<!-- <</Stencil::Block>> -->

<!-- Notes that may be helpful for anyone reviewing this PR -->
## Notes for your reviewers



<!-- <<Stencil::Block(custom)>> -->

<!-- <</Stencil::Block>> -->

tonymbugua-outreach commented:
@jaredallard You mentioned that you would like to have the resolvers implementation in a separate dir but I was running into the import cycle error. Any feedback for a workaround would be greatly appreciated.

jaredallard commented on pkg/app/githubVersionResolver.go:
```diff
@@ -0,0 +1,34 @@
+func (g *GitHubVersionResolver) GetLatestVersion(ctx context.Context, log logrus.FieldLogger, app *App) (string, error) {
+	// Find the latest version if not set, or resolve the provided version
+	if app.Version == "" {
+		if err := app.detectVersion(ctx); err != nil {
```
We should probably keep `app.Version` != "" logic before this, right? Since this should only be invoked if we need to get the latest version.

tonymbugua-outreach replied:
Hmm, i didnt change anything about the `app.Version` check from the [main branch](https://github.com/getoutreach/devenv/blob/main/internal/app/app.go#L154). Don't we want to get the latest version when there is no version specified?

tonymbugua-outreach replied:
not sure if i understand the ask here 😅

tonymbugua-outreach replied:
already done

jaredallard commented on pkg/app/githubVersionResolver.go:
```diff
@@ -0,0 +1,34 @@
+	} else {
+		// If we had a provided version, attempt to
+		// resolve the version in case it's a branch
+		if err := app.resolveVersion(ctx); err != nil {
```
I would suggest moving the `resolveVersion` code into this file, since it should be specific to this resolver now.

tonymbugua-outreach replied:
Moved the `resolveVersion` function to the githubVersionResolver.go file

jaredallard commented on pkg/app/githubVersionResolver.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+// GetLatestVersion returns the latest image version based on git repo
+func (g *GitHubVersionResolver) GetLatestVersion(ctx context.Context, log logrus.FieldLogger, app *App) (string, error) {
```
It looks like `*App` is the reason for the import cycle, I suggest just passing the appName in instead to deviate from the original design. This will prevent an import cycle and be much cleaner.

tonymbugua-outreach replied:
The function interface relies on the app struct to exist to reference some variables within it inorder for maestro to work. Not sure if this needs a rewrite then or we should leave it as is?

tonymbugua-outreach replied:
Fixed in call today

jaredallard commented on pkg/app/maestroVersionResolver.go:
```diff
@@ -0,0 +1,96 @@
+	"strings"
+
+	"github.com/getoutreach/gobox/pkg/box"
+	"github.com/getoutreach/maestro/pkg/mclient"
```
This is a private repo, so we can't clone in that library unfortunately. We'll need to either:

a) Pull it out and use it
b) Duplicate the logic

b seems like the better option for now since we can just make an HTTP request. We might want to make a ticket to do `a` in the long run.

tonymbugua-outreach replied:
Duplicated mclient into devenv and opened a [JIRA ticket](https://outreach-io.atlassian.net/browse/DT-3571) to track it

jaredallard commented on pkg/app/maestroVersionResolver.go:
```diff
@@ -0,0 +1,96 @@
+	// loop through the channels specified in config
+	// if the channel returns 404 retry
+
+	log = log.WithFields(logrus.Fields{
```
Would it be a good idea to change the logger being passed into these to set `app.name` and `app.versionResolver` automatically?

tonymbugua-outreach replied:
logger being passed in has the app.name and app.versionResolver set automatically now

jaredallard commented on pkg/app/maestroVersionResolver.go:
```diff
@@ -0,0 +1,96 @@
+	for _, channel := range versionResolver.Config.Maestro.Channels {
+		maestroVersion, err := maestroClient.FetchActionable(app.RepositoryName, channel)
+		if err != nil {
+			log.Info(fmt.Sprintf("unable to fetch actionable version for channel %s Err: %s", channel, err.Error()))
```
```suggestion
			log.WithField("channel", channel).WithError(err).Info("unable to fetch actionable version")
```

jaredallard replied:
I might be slightly wrong on the syntax here, so might not work as-is

tonymbugua-outreach replied:
Updated

jaredallard commented on pkg/app/resolver.go:
```diff
@@ -0,0 +1,45 @@
+		versionResolverType := VersionResolverType(versionResolverType)
+
+		// get config implementation based on type
+		config, err := app.getVersionResolverConfigFn(versionResolverType)
```
It might be a better idea to use a `New<Resolver>(b *box.Config)` function instead that passes in the box config. This would allow the package to determine what fields to grab off of the box config. This could be its config as well as potentially other fields it may need (e.g., vault for maestro)

tonymbugua-outreach replied:
Sounds good. I will work on trying to refactor this

tonymbugua-outreach replied:
fixed in call today

jaredallard commented on pkg/app/resolver_test.go:
```diff
@@ -0,0 +1,261 @@
+package app
```
love to see the test coverage! ❤️

jaredallard commented on pkg/app/versionResolver.go:
```diff
@@ -0,0 +1,39 @@
+package app
```
General comment for all the files, we're missing the copyright and some exported types don't have comments. It'd be great to get that added!

tonymbugua-outreach replied:
added to all files

jaredallard commented on pkg/app/mockVersionResolver.go:
```diff
@@ -0,0 +1,64 @@
+	"github.com/sirupsen/logrus"
+)
+
+type MockVersionResolver struct {
```
Slight nit, I like to build mock providers that store data instead of explicit `ReturnsErr` options. This makes it a little more natural to use (since If you operate on what you think it'd return, it's easier to determine if it should return an error or not).

Here's an example: https://github.com/getoutreach/argocd/blob/master/maestro-dv-syncer/internal/argocd/argocdtest/argocdtest.go

tonymbugua-outreach replied:
Hmm, I spent a few trying to figure out what to store as data but the only data that gets returned is the version (string) and error if its unable to get the version from that version resolver. Maybe we can chat about this more in the meeting I added for Monday next week

tonymbugua-outreach replied:
```
type VersionResolver interface {
	GetLatestVersion(context.Context, logrus.FieldLogger, *App) (string, error)
}
```

jaredallard commented on pkg/app/versionResolver.go:
```diff
@@ -0,0 +1,39 @@
+
+// StringToVersionResolverConfigMap converts a VersionResolverType to a
+// VersionResolver config if found. Returns an error if not found
+var StringToVersionResolverConfigMap = map[string]VersionResolver{
```
I would suggest using `VersionResolverType` as the key type here and converting the user provided string. Also, `Type` at the end of it feels not idiomatic since it includes the type in the name, I suggest just going with `VersionResolverName`

```
config, found := StringToVersionResolverConfigMap[VersionResolverName(key)]
```

tonymbugua-outreach replied:
updated map to use `VersionResolverName` as the key and changed datatpe from VersionResolverType to VersionResolverName

jaredallard commented on internal/app/resolvers/github/github.go:
```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: This file has the package github.
+
+// Package github handles the github implementation for VersionResolver.
+package github
```
```suggestion
// Copyright 2023 Outreach Corporation. All Rights Reserved.

// Description: See package comment.

// Package github implements the VersionResolver interface using Github as the backing
// version store. Both commits and tags are supported as a source of truth for the latest
// version of an application.
//
// For tags, semantic version constructs are used. The latest semantic version is used.
// For commits, the latest commit in the default branch (determined through the Github API)
// is used.
//
// See`detectVersion` for more information about the behaviour and how to select
// commits and tags for cases where both may be in use.
package github
```

jaredallard commented on internal/app/resolvers/github/github.go:
```diff
@@ -0,0 +1,91 @@
+	"golang.org/x/oauth2"
+)
+
+type VersionResolver struct {
+	boxConfig *box.Config
+}
+
+func New(b *box.Config) types.VersionResolver {
+	return &VersionResolver{
```
```suggestion
// VersionResolver implements the VersionResolver interface.
type VersionResolver struct {
	boxConfig *box.Config
}

// New returns a new Github VersionResolver implementation
func New(b *box.Config) types.VersionResolver {
	return &VersionResolver{
```

jaredallard commented on internal/app/resolvers/github/github.go:
```diff
@@ -0,0 +1,91 @@
+)
+
+type VersionResolver struct {
+	boxConfig *box.Config
```
nit: It's more idiomatic to use `b` instead of `boxConfig` here

jaredallard replied:
(`box` is also acceptable)

jaredallard commented on internal/app/resolvers/maestro/maestro.go:
```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: This file has the package maestro.
+
+// Package maestro contains the implementation of maestro for VersionResolver
```
```suggestion
// Copyright 2023 Outreach Corporation. All Rights Reserved.

// Description: See the package comment.

// Package maestro contains a VersionResolver implementation using Maestro, and internal
// service in use at Outreach, as the source of truth for the latest version of applications.
```

jaredallard commented on internal/app/resolvers/maestro/maestro.go:
```diff
@@ -0,0 +1,96 @@
+	boxConfig *box.Config
+}
+
+func New(b *box.Config) types.VersionResolver {
```
```suggestion
// New returns a new Maestro VersionResolver
func New(b *box.Config) types.VersionResolver {
```

jaredallard commented on internal/app/resolvers/mock/mock.go:
```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: This file implements mock client.
+
+// Package mock contains the implementation of mock for VersionResolver
```
```suggestion
// Copyright 2023 Outreach Corporation. All Rights Reserved.

// Description: See package comment.

// Package mock implements a VersionResolver implementation for use in testing.
// This implementation returns static data only and is not meant to be used outside
// of testing environments.
```

jaredallard commented on internal/app/resolvers/types/types.go:
```diff
@@ -0,0 +1,23 @@
+type VersionResolver interface {
+	GetLatestVersion(context.Context, logrus.FieldLogger, string, string) (string, error)
+}
+
+type VersionResolverCreator func(*box.Config) VersionResolver
```
```suggestion

// VersionResolverCreator is the standard format of a `New` function
// that return a new instance of an implementation of the `VersionResolver`
// interface.
type VersionResolverCreator func(*box.Config) VersionResolver
```

jaredallard commented on internal/app/versionResolver.go:
This file feels like belongs in the `resolvers` folder instead in potentially something like `resolvers.go`. It'll also compose better there.

tonymbugua-outreach replied:
Tried to move it but the import cycle error comes back again. App is importing the resolvers package and the resolver_test.go file is importing the app package 😢

jaredallard commented on internal/app/versionResolver.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// StringToVersionResolverConfigMap converts a VersionResolverType to a
+// VersionResolver config if found. Returns an error if not found
```
```suggestion
// VersionResolver config if found.
```

Maps don't return errors if a key doesn't exist, so slight adjustment there.

jaredallard commented on pkg/embed/manifests/pre-restore/cert-manager.yaml:
Accidental commit?

tonymbugua-outreach replied:
reverted

jaredallard commented on internal/app/resolver.go:
```diff
@@ -0,0 +1,57 @@
+		versionResolverType := types.VersionResolverName(versionResolverType)
+
+		// get config implementation based on type
+		config, err := app.getVersionResolverConfigFn(versionResolverType)
```
This isn't really the config, at this point, since it's the resolver itself. I suggest renaming this to `resolver` and updating the names. This should just be `getVersionResolverByName`.

jaredallard commented on internal/app/resolver.go:
```diff
@@ -0,0 +1,57 @@
+	versionResolvers := app.box.DeveloperEnvironmentConfig.VersionResolvers
+
+	// loop through all the enabled versionResolver providers
+	for _, versionResolverType := range versionResolvers.Enabled {
```
A lot of the variable names in this function are long. I'd suggest dropping redundant terms where it makes sense. For example, this function is named `HandleVersionResolver`. We can probably drop `versionResolver` from being a prefix on most of these.

jaredallard reviewed (commented):
Left some comments.

Overall, we're close! I'd like to see the maestro client ripped out, though; it's currently leaking many implementation details about a private service. The original design called for just using a standard Golang HTTP client since we only make a single call. Let's continue doing that. It would also reduce the LOC introduced in this PR a lot.

tonymbugua-outreach commented:
> Left some comments.
> 
> Overall, we're close! I'd like to see the maestro client ripped out, though; it's currently leaking many implementation details about a private service. The original design called for just using a standard Golang HTTP client since we only make a single call. Let's continue doing that. It would also reduce the LOC introduced in this PR a lot.

Sounds good! Refactored and removed the client. Changed to use API

jaredallard commented on internal/app/resolvers/maestro/client/client.go:
You should be able to replace this file with `maestroClient.go`, right?

tonymbugua-outreach replied:
This was the [error](https://app.circleci.com/pipelines/github/getoutreach/devenv/2096/workflows/f8125f5b-5939-4869-94f2-99c305da446e/jobs/5962) coming out of circleCI

jaredallard replied:
Ah, yes, I meant rename `maestroClient.go` to `client.go` because it's already in a descriptive path

tonymbugua-outreach replied:
oh i see. Yes

jaredallard commented on internal/app/resolvers/maestro/client/maestroClient.go:
```diff
@@ -0,0 +1,75 @@
+	}
+	defer response.Body.Close()
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var version Version
+	err = json.Unmarshal(body, &version)
+	if err != nil {
+		return nil, err
+	}
```
nit: You can avoid reading the response into memory here with `json.NewDecoder(resp.Body).Decode(&version)`

jaredallard commented on internal/app/resolvers/maestro/client/version.go:
```diff
@@ -0,0 +1,17 @@
+
+// Version is the struct payload returned from maestro actionableVersion
+type Version struct {
+	// we are making the explicit decision that 0 is not a valid id and want to omit it from json
```
```suggestion
```

I don't feel like this comment adds much value
